### PR TITLE
notifications

### PR DESCRIPTION
**Pull request by m-mcgowan**
# Problem Context

We want to show notifications when a new node is paired, and perhaps later when a new gateway is detected, and for other reasons - notifications have general application. 

Notifications are stored persistently in the datastore until they are dismissed, since the serverless deployment is stateless.

There is a known issue where dismissing a notification causes a client side warning 
```
next-dev.js?3515:32 Warning: findDOMNode is deprecated in StrictMode. findDOMNode was passed an instance of DomWrapper which is inside StrictMode. Instead, add a ref directly to the element you want to reference. Learn more about using refs safely here: https://reactjs.org/link/strict-mode-find-node
    at div
```    

## Screenshot (if applicable)

image

## Testing

Testing with the simulator:
 * Have the local sparrow app setup against a notehub project, with an active route to the local tunnel, and use `hub.set` to point the simulator to that project.
 * In the simulator, send this request (one or more times.) The NodeID's can be real or made-up
 ```
{"cmd":"hub.log","method":"sensor-provision","text":"20323746323650050018dddd","sync":true}
```
* A few seconds later, the webapp will show the notification that a new node was paired.
  * Clicking the 'x' closes the notification with no further action. Refreshing the page shows that the notification has been removed. 
  * Clicking the "Setup" button opens the settings page for the newly paired node and removes the notification. 
* All notifications are displayed on each page of the app.  
* Notifications not yet dismissed survive restarts of the app.

### CODE REVIEW

File: __tests__/src/adapters/nextjs-sparrow/NextJsUrlManager.test.ts [hunk 1]
```diff
@@ -3,10 +3,65 @@
...
   it("should return correct endpoint url", () => {
     expect(
       NextJsUrlManager.nodeNameUpdate("gfoo", "sbar")
     ).toMatchInlineSnapshot(`"/api/gateway/gfoo/node/sbar/config"`);
   });
 });
+
+describe(NextJsUrlManager.presentNotifications, () => {
+  it("returns the base API endpoint when no ids are present", () => {
+    expect(NextJsUrlManager.presentNotifications()).toMatchInlineSnapshot(
```
Comment (paigen11) [inline __tests__/src/adapters/nextjs-sparrow/NextJsUrlManager.test.ts:19]: TIL `toMatchInlineSnapshot` - I didn't know that was a method in Jest. Very handy

File: __tests__/src/adapters/nextjs-sparrow/NextJsUrlManager.test.ts [hunk 2]
```diff
@@ -3,10 +3,65 @@
...
+      `"/api/notifications?present=1"`
+    );
+  });
+
+  it("returns a signle ID as an 'id' query parameter", () => {
```
Comment (paigen11) [inline __tests__/src/adapters/nextjs-sparrow/NextJsUrlManager.test.ts:24]:
```suggestion
  it("returns a single ID as an 'id' query parameter", () => {
```

File: __tests__/src/adapters/nextjs-sparrow/NextJsUrlManager.test.ts [hunk 3]
```diff
@@ -3,10 +3,65 @@
...
+      `"/api/notifications?id=foo"`
+    );
+  });
+
+  it("returns multiple ID as multiple 'id' query parameters", () => {
+    expect(
+      NextJsUrlManager.notifications("foo", "bar", "baz")
+    ).toMatchInlineSnapshot(`"/api/notifications?id=foo&id=bar&id=baz"`);
+  });
+});
```
Comment (paigen11) [inline __tests__/src/adapters/nextjs-sparrow/NextJsUrlManager.test.ts:66]: This looks like a duplicate of the tests above?

File: src/api-client/notification.ts [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+export async function fetchNotifications() {
+  const endpoint = services().getUrlManager().notifications();
+  const response: AxiosResponse = await axios.get(endpoint);
+  return response.data;
+}
+
+
+
+const DEFAULT = { removeNotification };
+export default DEFAULT;
```
Comment (paigen11) [inline src/api-client/notification.ts:38]: What's the purpose of this code? I don't believe I've seen this implementation before

File: src/components/elements/Notifications.tsx [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+import styles from "../../styles/Home.module.scss";
+import notificationsStyles from "../../styles/Notifications.module.scss";
+
+async function nodePairedAction(n: NodePairedWithGatewayAppNotification, router: NextRouter) {
+  console.log("redirecting", n);
+  try {
+    await removeNotification(n.id);
+  }
+  catch (e) {
+    console.log(e);
```
Comment (paigen11) [inline src/components/elements/Notifications.tsx:21]: Perhaps switch this to a `console.error()` for the `catch` statement

File: src/pages/api/notifications.ts [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+function asNotification(o: object): Notification {
+  if (isNotification(o)) return o as Notification;
+  else throw invalidArgumentError();
+}
+
+function asNotifications(body: any): Notification[] {
+  const { notifications } = body;
+  if (Array.isArray(notifications)) {
+    throw invalidArgumentError();
+  }
```
Comment (paigen11) [inline src/pages/api/notifications.ts:28]: Should this be `if (!Array.isArray(notifications)) {` ?

File: src/pages/index.tsx [hunk 7]
```diff
@@ -43,11 +44,16 @@ const Home: NextPage<HomeData> = ({ gatewayNodeData, err }) => {
       {err ? (
         <h2
           className={styles.errorMessage}
+          // life in the fast lane...
+          // eslint-disable-next-line react/no-danger
```
Comment (paigen11) [inline src/pages/index.tsx:48]: 😎 🏎️

File: src/services/notehub/SparrowEvents.ts [hunk 8]
```diff
@@ -5,6 +5,11 @@ import NotehubRoutedEvent, {
   NotehubRoutedEventLocationFields,
 } from "./models/NotehubRoutedEvent";
 
+export const _health = {
+  qo: "_health.qo",
+  SENSOR_PROVISION: "sensor-provision",
```
Comment (paigen11) [inline src/services/notehub/SparrowEvents.ts:10]: Good use of constants

**Review by paigen11 — APPROVED** [review]
This looks really great. I'd love to see a demo of the notifications in action if you have time to show us!

**Comment by m-mcgowan** [comment]
[sparrow reference web app: Notify users when a node has been paired.](https://trello.com/c/3S8LfnCo/34-sparrow-reference-web-app-notify-users-when-a-node-has-been-paired)

File: __tests__/src/adapters/nextjs-sparrow/NextJsUrlManager.test.ts (hunk 3 above)
Comment (m-mcgowan) [reply __tests__/src/adapters/nextjs-sparrow/NextJsUrlManager.test.ts:66]: `notifications` vs `presentNotifications`. The outcomes are slightly different (a `present=1` query param.)

File: src/api-client/notification.ts (hunk 4 above)
Comment (m-mcgowan) [reply src/api-client/notification.ts:38]: I'm not entirely sure either, I was copying existing code, but I'd prefer `export default`.

File: src/components/elements/Notifications.tsx [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+    .nodeSettings(n.gateway.uid, n.node.nodeId);
+  console.log("redirecting to", url);
+
+  await router.push(url);
+}
+
+function renderNodePairedNotification(n: NodePairedWithGatewayAppNotification, router: NextRouter) {
+
+  const note = <span>A new node was recently paired.</span>;
+  //const details = <span>Node ID: <span>{n.node.nodeId}</span></span>
```
[the remark this answers is not in the document]
Comment (m-mcgowan) [reply src/components/elements/Notifications.tsx:36]: Sure - I'd like us to discuss what goes in the notification message. We may want to include other details.

File: __tests__/src/adapters/nextjs-sparrow/NextJsUrlManager.test.ts (hunk 3 above)
Comment (m-mcgowan) [reply __tests__/src/adapters/nextjs-sparrow/NextJsUrlManager.test.ts:66]: Ah, I see I'd missed committing the latest changes. They're fixed up now, but not when you first reviewed. Good eye!

File: src/api-client/notification.ts (hunk 4 above)
Comment (m-mcgowan) [reply src/api-client/notification.ts:38]: In our huddle, we found out it's a way of avoiding a nameless default export.

File: src/pages/api/notifications.ts (hunk 6 above)
Comment (m-mcgowan) [reply src/pages/api/notifications.ts]: Code removed :-)

**Review by CarltonHenderson — APPROVED** [review]

**Pull request merged**